src/libraries/QueryResponse.sol

```diff
@@ -0,0 +1,590 @@
+      revert WrongQueryType(pcr.queryType, QueryType.SOLANA_PDA);
+
+    uint reqOff;
+    uint respOff;
+
+    (ret.requestCommitment,      reqOff) = pcr.request.sliceUint32PrefixedUnchecked(reqOff);
+    (ret.requestMinContextSlot,  reqOff) = pcr.request.asUint64Unchecked(reqOff);
+    (ret.requestDataSliceOffset, reqOff) = pcr.request.asUint64Unchecked(reqOff);
+    (ret.requestDataSliceLength, reqOff) = pcr.request.asUint64Unchecked(reqOff);
+
+    uint8 numPdas;
+    (numPdas, reqOff) = pcr.request.asUint8Unchecked(reqOff);
+
+    (ret.slotNumber, respOff) = pcr.response.asUint64Unchecked(respOff);
+    (ret.blockTime,  respOff) = pcr.response.asUint64Unchecked(respOff);
+    (ret.blockHash,  respOff) = pcr.response.asBytes32Unchecked(respOff);
+
+    uint8 respNumResults;
+    (respNumResults, respOff) = pcr.response.asUint8Unchecked(respOff);
+    if (respNumResults != numPdas)
+      revert UnexpectedNumberOfResults();
+
+    ret.results = new SolanaPdaResult[](numPdas);
+
+    //walk through the call inputs and outputs in lock step.
+    for (uint i; i < numPdas; ++i) {
+      (ret.results[i].programId, reqOff) = pcr.request.asBytes32Unchecked(reqOff);
+
+      uint8 reqNumSeeds;
+      (reqNumSeeds, reqOff) = pcr.request.asUint8Unchecked(reqOff);
+      ret.results[i].seeds = new bytes[](reqNumSeeds);
+      for (uint s; s < reqNumSeeds; ++s)
+        (ret.results[i].seeds[s], reqOff) = pcr.request.sliceUint32PrefixedUnchecked(reqOff);
+
+      (ret.results[i].account,    respOff) = pcr.response.asBytes32Unchecked(respOff);
+      (ret.results[i].bump,       respOff) = pcr.response.asUint8Unchecked(respOff);
+      (ret.results[i].lamports,   respOff) = pcr.response.asUint64Unchecked(respOff);
+      (ret.results[i].rentEpoch,  respOff) = pcr.response.asUint64Unchecked(respOff);
+      (ret.results[i].executable, respOff) = pcr.response.asBoolUnchecked(respOff);
+      (ret.results[i].owner,      respOff) = pcr.response.asBytes32Unchecked(respOff);
+      (ret.results[i].data,       respOff) = pcr.response.sliceUint32PrefixedUnchecked(respOff);
+    }
+
+    _checkLength(pcr.request, reqOff);
+    _checkLength(pcr.response, respOff);
+  }}
+
+  function validateBlockTime(
+    uint64 blockTimeInMicroSeconds,
+    uint256 minBlockTimeInSeconds
+  ) internal pure {
+    uint256 blockTimeInSeconds = blockTimeInMicroSeconds / MICROSECONDS_PER_SECOND; // Rounds down
+
+    if (blockTimeInSeconds < minBlockTimeInSeconds)
+      revert StaleBlockTime();
+  }
+
+  function validateBlockNum(uint64 blockNum, uint256 minBlockNum) internal pure {
+    if (blockNum < minBlockNum)
+      revert StaleBlockNum();
+  }
+
+  function validateChainId(
+    uint16 chainId,
+    uint16[] memory validChainIds
+  ) internal pure { unchecked {
+    uint len = validChainIds.length;
+    for (uint i; i < len; ++i)
+      if (chainId == validChainIds[i])
+        return;
+
+    revert InvalidChainId();
+  }}
+
+  function validateEthCallRecord(
```

Comment:
An integrator recently elected not to use this method as they simply wanted to validate that the response came from a single address with a single function selector; which in hindsight is probably the most common use case and the "catch-all" approach of this method that I originally added might not be as useful as I thought.

Wondering if we should keep this AND add methods to check a single address/selector, OR replace this method (and the associated functions it calls into) altogether?

Reply:
Ok cool, yeah I think both comments fall into the same bucket. Sounds like a plan to me, I'm sure there's some other feedback floating around that I'm not aware of that (the royal) we could collate and prioritise.